swiftwinrt/Resources/Support/Error.swift

```diff
@@ -167,8 +168,109 @@ public struct Error : Swift.Error, CustomStringConvertible {
     self.description = getErrorDescription(expecting: hr) ?? hrToString(hr)
     self.hr = hr
   }
+
+  fileprivate init(hr: HRESULT, file: StaticString, line: UInt) {
+    self.description = "\(file): \(line) - \(getErrorDescription(expecting: hr) ?? hrToString(hr))"
+    self.hr = hr
+  }
 }
 
 public func failWith(err: HRESULT) -> HRESULT {
   return err
 }
+
+// # Mark - WinRT Error Handling
+// Because all WinRT methods return HRESULTs and every property get/set or event registration returns an HRESULT,
+// we'd have to mark all WinRT API calls with "throws" in order to properly handle errors. However, this makes
+// the API less ergonomic to use, and since most of these calls are due to improper usage, try! is often the
+// write outcome for the developer to use. So, instead of marking all of these methods with "throws", we provide
+// the ability to handle the error through the `withFailingCall` method below
+
+/// Handle any failing WinRT API call that is not marked with `throws`. This call can be used to catch exceptions
+/// from non-throwing WinRT calls such as properties or event registrations in the abnormal case where handling these
+/// failures is required.
+///
+/// Example Usage:
+///    let curentPackage = try? withFailingCall(Package.current)
+public func withFailingCall<Result>(_ call: @autoclosure () -> Result) throws -> Result {
+  let error = getLastError()
+  defer { clearLastError() }
+  let result = call()
+  if let error = error.error {
+    throw error
+  }
+  return result
+}
+
+// # MARK - WinRT Error Handling implementation
+// The code generator wraps all non-throwing WinRT API calls in `_tryWinRT` to handle the error and set the last error.
+// There can only be one last error at a time, so the last error is cleared after it is read and protected against
+// potential access from different threads.
+
+@_spi(WinRTInternal)
+public func _tryWinRT<Result>(
+  _ defaultValue: Result,
+  _ call: @autoclosure () throws -> Result,
+    file: StaticString = #fileID,
+    function: StaticString = #function,
+    line: UInt = #line) -> Result {
+    do {
+      return try call()
+    } catch {
+      let error = error as! Error
+      setLastError(Error(hr: error.hr, file: file, line:line))
+      // This "default" value is returned in the case of an error, but it is
+      // never returned to the caller because the method is going to throw
+      return defaultValue
+    }
+}
+
+@_spi(WinRTInternal)
+public func _tryWinRT(
+  _ call: @autoclosure () throws -> (),
+    file: StaticString = #fileID,
+    function: StaticString = #function,
+    line: UInt = #line) {
+    do {
+      try call()
+    } catch {
+      let error = error as! Error
+      setLastError(Error(hr: error.hr, file: file, line:line))
+    }
+}
+
+fileprivate var getErrorLock = NSLock()
+fileprivate var checkErrorLock = NSLock()
+fileprivate var lastError: LastErrorHolder?
```

Comment:
This would need to be thread-local. Concurrent calls should have independent last errors to not step on each other's toes.

Reply:
thread local would prevent the need for locks, but the last error is only used in the scope of the `withFailingCall` method, so the other thread will block waiting to get access to the last error. at the end of `withFailingCall` the error is cleared